File: src/infrastructure/tools/knowledge-base/list-knowledge-bases.ts
/**
 * list_knowledge_bases — 列出所有可用知识库
 *
 * 辅助工具，帮助 AI 了解有哪些知识库可供检索和写入。
 */
import { toolRegistry } from '@/domain/tools'
import type { ToolExecutor, ToolResult } from '@/domain/tools/types'
import { t } from '@/ui/i18n'
import { ragService } from '@/services/rag-service'

toolRegistry.register(
  {
    name: 'list_knowledge_bases',
    label: t('列出知识库'),
    description:
      'List all available knowledge bases. ' +
      'Returns the name, description, ID, and document count for each knowledge base. ' +
      'Use this tool to discover which knowledge bases are available before searching or writing. ' +
      'The ID field is required for search_knowledge_base and write_to_knowledge_base tools.',
    parameters: {
      type: 'object',
      properties: {},
      required: [],
    },
  },
  (async (_args: Record<string, any>, _ctx: any): Promise<ToolResult> => {
    try {
      const kbs = await ragService.listKnowledgeBases()

      if (!kbs || kbs.length === 0) {
        return {
          content:
            'No knowledge bases found. ' +
            'Create a knowledge base first and upload documents to it, ' +
            'or use write_to_knowledge_base to create new content.',
          uiData: { length: 0 },
        }
      }

      // 为每个知识库获取文档标题（最多 20 个）
      const kbDocPromises = kbs.map(async (kb) => {
        let docs: Array<{ file_name: string; id: string }> = []
        try {
          const result = await ragService.listDocuments(kb.id)
          docs = result.slice(0, 20).map((d) => ({ file_name: d.file_name, id: d.id }))
        } catch {
          // 单个知识库查询失败不影响其他
        }
        return { kb, docs }
      })

      const kbDocs = await Promise.all(kbDocPromises)

      const lines: string[] = []
      lines.push(`📚 Available Knowledge Bases (${kbs.length} total):`)
      lines.push('')

      for (let i = 0; i < kbDocs.length; i++) {
        const { kb, docs } = kbDocs[i]
        lines.push(`[${i + 1}] ${kb.name}`)
        lines.push(`    ID: ${kb.id}`)
        lines.push(`    Description: ${kb.description || 'No description'}`)
        lines.push(`    Documents: ${kb.document_count}`)
        lines.push(`    Chunks: ${kb.chunk_count}`)

        if (docs.length > 0) {
          const docLines = docs.map((d, idx) => `      ${idx + 1}. ${d.file_name}  (ID: ${d.id})`)
          lines.push(`    Document titles (showing ${docs.length} of ${kb.document_count}):`)
          lines.push(docLines.join('\n'))
        }
        lines.push('')
      }

      lines.push('---')
      lines.push(
        'Use list_knowledge_base_documents with the knowledge_base_id to see all documents and their IDs.',
      )
      lines.push(
        'Use search_knowledge_base with the knowledge_base_id to search within a specific knowledge base.',
      )
      lines.push(
        'Use write_to_knowledge_base with the knowledge_base_id to save new content.',
      )

      return {
        content: lines.join('\n'),
        uiData: {
          length: kbs.length,
          knowledgeBases: kbs.map((kb) => ({
            id: kb.id,
            name: kb.name,
            description: kb.description,
            documentCount: kb.document_count,
          })),
        },
      }
    } catch (error: any) {
      return {
        content: `Error listing knowledge bases: ${error.message || String(error)}`,
      }
    }
  }) as ToolExecutor,
)

File: src-tauri/virlen-core/src/agent/native_tools/knowledge_base/list_knowledge_bases.rs
//! `list_knowledge_bases` 工具（原生）— 列出全部知识库（含文档数/分块数与部分文档标题）。

use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};

use super::common::rag_service;

pub(crate) async fn list_knowledge_bases_tool(
    _ctx: &NativeToolCtx<'_>,
    _args: &Value,
) -> Result<NativeToolOutcome, String> {
    let service = rag_service()?;
    let kbs = tokio::task::spawn_blocking(move || service.list_knowledge_bases())
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Error listing knowledge bases: {}", e))?;

    if kbs.is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: "No knowledge bases found. Create a knowledge base first and upload documents to it, or use write_to_knowledge_base to create new content.".to_string(),
            ui_data: Some(json!({ "length": 0 })),
        });
    }

    let mut lines = vec![format!("📚 Available Knowledge Bases ({} total):", kbs.len()), String::new()];
    let mut ui_kbs: Vec<Value> = Vec::new();

    for (i, kb) in kbs.iter().enumerate() {
        // 每个知识库列出最多 20 个文档标题
        let docs = service.list_documents(&kb.id).unwrap_or_default();
        let doc_titles: Vec<String> = docs.iter().take(20).map(|d| d.file_name.clone()).collect();

        lines.push(format!("[{}] {}", i + 1, kb.name));
        lines.push(format!("    ID: {}", kb.id));
        lines.push(format!("    Description: {}", if kb.description.is_empty() { "No description".to_string() } else { kb.description.clone() }));
        lines.push(format!("    Documents: {}", kb.document_count));
        lines.push(format!("    Chunks: {}", kb.chunk_count));
        if !doc_titles.is_empty() {
            lines.push(format!("    Document titles (showing {} of {}):", doc_titles.len(), kb.document_count));
            for (idx, title) in doc_titles.iter().enumerate() {
                lines.push(format!("      {}. {}", idx + 1, title));
            }
        }
        lines.push(String::new());

        ui_kbs.push(json!({
            "id": kb.id,
            "name": kb.name,
            "description": kb.description,
            "documentCount": kb.document_count,
        }));
    }

    lines.push("---".to_string());
    lines.push("Use list_knowledge_base_documents with the knowledge_base_id to see all documents and their IDs.".to_string());
    lines.push("Use search_knowledge_base with the knowledge_base_id to search within a specific knowledge base.".to_string());
    lines.push("Use write_to_knowledge_base with the knowledge_base_id to save new content.".to_string());

    Ok(NativeToolOutcome::Value {
        content: lines.join("\n"),
        ui_data: Some(json!({ "length": kbs.len(), "knowledgeBases": ui_kbs })),
    })
}
